src/explore-education-statistics-common/src/modules/table-tool/utils/__tests__/applyTableHeadersOrder.test.ts

```diff
@@ -0,0 +1,1542 @@
+            testCategoryFilters1[4],
+          ],
+        ],
+      };
+      const expected: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [
+          [
+            testCategoryFilters1[2],
+            testCategoryFilters1[0],
+            testCategoryFilters1[3],
+            testCategoryFilters1[4],
+          ],
+        ],
+      };
+      const result = applyTableHeadersOrder({
+        reorderedTableHeaders: testReorderedHeadersForRowsInRowGroups,
+        updatedTableHeaders: testUpdatedTableHeaders,
+      });
+      expect(result).toEqual(expected);
+    });
+  });
+
+  describe('rowGroups', () => {
+    const testReorderedHeadersForRowGroups: TableHeadersConfig = {
+      ...testHeaders,
+      rowGroups: [
+        testCategoryFilters2,
+        testLocationFilters,
+        testCategoryFilters1,
+      ],
+      rows: testIndicators,
+    };
+
+    test('retains the order of reordered rowGroups', () => {
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [
+          testCategoryFilters1,
+          testCategoryFilters2,
+          testLocationFilters,
+        ],
+        rows: testIndicators,
+      };
+      const result = applyTableHeadersOrder({
+        reorderedTableHeaders: testReorderedHeadersForRowGroups,
+        updatedTableHeaders: testUpdatedTableHeaders,
+      });
+      expect(result).toEqual(testReorderedHeadersForRowGroups);
+    });
+
+    test('retains the order when rowGroups have been removed', () => {
+      const testReorderedHeaders: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [
+          testLocationFilters,
+          testCategoryFilters1,
+          testCategoryFilters2,
+        ],
+        rows: testIndicators,
+      };
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [testCategoryFilters2, testLocationFilters],
+        rows: testIndicators,
+      };
+      const expected: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [testLocationFilters, testCategoryFilters2],
+        rows: testIndicators,
+      };
+      const result = applyTableHeadersOrder({
+        reorderedTableHeaders: testReorderedHeaders,
+        updatedTableHeaders: testUpdatedTableHeaders,
+      });
+
+      expect(result).toEqual(expected);
+    });
+
+    test('adds new rowGroups after reordered rowGroups', () => {
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [
+          testCategoryFilters1,
+          testCategoryFilters2,
+          testLocationFilters,
+          testCategoryFilters3,
+          testCategoryFilters4,
+        ],
+        rows: testIndicators,
+      };
+      const expected: TableHeadersConfig = {
+        ...testHeaders,
+        rowGroups: [
+          testCategoryFilters2,
+          testLocationFilters,
+          testCategoryFilters1,
+          testIndicators,
```

Comment:
From a user's point of view there are just row groups, `rows` is an internal term for the last of the row groups.

If they've reordered the row groups it might, as you say, be because they want filter X to be the last of the groups. But it also might be because they want filter Y to be the first of the groups. There's no way to infer their intent. So I've put any new groups after the existing ones, which is arbitrary, but feels better to me than putting them at the start or before the last one (between rows and rowGroups).